trieste/acquisition/function.py

```diff
@@ -524,3 +527,194 @@ def prepare_acquisition_function(
+    def prepare_acquisition_function(
+        self, dataset: Dataset, model: ProbabilisticModel
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: The data to use to build the acquisition function.
+        :param model: The model over the specified ``dataset``.
+        :return: An acquisition function.
+        """
+
+
+def predict_independent_f_samples_with_reparametrisation_trick(
+        model: ProbabilisticModel, at: QueryPoints, eps: tf.Tensor
+) -> tf.Tensor:
+    """
+    Returns independent samples according to the reparametrization trick.
+    The sample size S is determined by eps,
+    N is the number of query points and L is the number of latent processes.
+    :param model:
+    :param at: N query points [N, D]
+    :param eps: a [S, L] tensor of N(0, 1) i.i.d. samples
+    :return: a [S, N, L] tensor
+    """
+    assert len(at.shape) == 2
+    mean, cov = model.predict(at)   # both [N, L]
+    return mean[None, ...] + tf.math.sqrt(cov)[None, ...] * eps[:, None, :]  # [S, N, L]
+
+
+def predict_batch_f_samples_with_reparametrisation_trick(
+        model: ProbabilisticModel, at: QueryPoints, eps: tf.Tensor
+) -> tf.Tensor:
+    """
+    Returns batch-correlated samples according to the reparametrization trick.
+    The sample size S is determined by eps,
+    the batch size B must be compatible between eps and at,
+    N is the number of batches of query points
+    and L is the number of latent processes.
+    :param model:
+    :param at: N batches of query points [N, B, D]
+    :param eps: a [S, B, L] tensor of N(0, 1) i.i.d. samples
+    :return: a [S, N, B, L] tensor
+    """
+
+    assert at.shape[1] == eps.shape[1]
+
+    mean, cov = model.predict(at, full_cov=True)
+    # mean: [N, B, L]
+    # cov: [N, L, B, B]
+    mean_for_sample = tf.linalg.adjoint(mean)  # [N, L, B]
+    mean_shape = tf.shape(mean_for_sample)
+    B = mean_shape[-1]
+
+    jittermat = (
+            tf.eye(B, batch_shape=mean_shape[:-1], dtype=default_float()) * default_jitter()
+    )  # [N, L, B, B]
+
+    eps = tf.transpose(eps, [2, 1, 0])  # [S, B, L] -> [L, B, S]
+
+    chol = tf.linalg.cholesky(cov + jittermat)  # [N, L, B, B]
+    samples = mean_for_sample[..., None] + tf.linalg.matmul(chol, eps)  # [N, L, B, S]
+    return tf.transpose(samples, [3, 0, 2, 1])
+
+
+class MonteCarloExpectedImprovement(SingleModelAcquisitionBuilder):
+    """
+    Builder for the Monte_carlo based expected improvement.
+    """
+
+    def __init__(self, num_samples: int):
+        super().__init__()
+        self.num_samples = num_samples
+
+    def prepare_acquisition_function(
+            self, dataset: Dataset, model: GPflowPredictor
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: The data from the observer.
+        :param model: The model over the specified ``dataset``.
+        :return: The expected improvement function.
+        """
+        eps_shape = [self.num_samples, model.model.num_latent_gps]
```

Comment:
in general, `model.model` is not possible, because the acquisition functionality uses `ProbabilisticModel`s which don't have underlying model. If we're working with vector targets, we can get the number of latent gps from e.g. the predictive mean. I haven't thought about how we'll achieve higher rank points across the toolbox.

Reply:
ah, ok, I've scrolled down a bit to "BatchMonteCarloExpectedImprovement", which partially answers my question. So my next question is then: what does "MonteCarloExpectedImprovement" that "ExpectedImprovement" doesn't?